Skuenzer/readme (#7)
unikraft/app-elfloader: README.md

# Unikraft ELF Loader

## Building the ELF Loader unikernel

`elfloader` can be built using the provided `Makefile`.

> Make sure the source code for [Unikraft](https://github.com/unikraft/unikraft), [libelf](https://github.com/unikraft/lib-libelf) and [lwip](https://github.com/unikraft/lib-lwip) is available at the paths referenced by the `UK_ROOT` and `UK_LIBS` variables respectively.

Reviewer: ```suggestion
Make sure the source code for [Unikraft](https://github.com/unikraft/unikraft), [libelf](https://github.com/unikraft/lib-libelf) and [lwip](https://github.com/unikraft/lib-lwip) is available at the paths referenced by the `UK_ROOT` and `UK_LIBS` Makefile variables, respectively.
```

Author: :+1:

> Execute the `make menuconfig` command to enter the configuration dialog, select a target platform (e.g. "KVM"), and exit to generate the unikernel configuration.

Reviewer: ```suggestion
Execute the `make menuconfig` command to enter the configuration dialog, select a target platform (e.g. `KVM`), and exit to generate the unikernel configuration.
```

Author: :+1:

> Executing `make` a second time triggers a build of the unikernel for the selected platform(s).

Reviewer: This is not enough. There are several options that must be enabled for a proper build, that includes networking, virtual memory and others:
- The `vfscore` configuration with `fs0` is mentioned below.
- `ukmmap` must be disabled, and then `ukvmem` and `posix-mmap` must be enabled.
- Paging must be enabled in `Platform Configuration` -> `Platform Interface Options` -> `Virtual memory API`.
- `Virtio PCI device support` must be enabled in `Platform Configuration` -> `KVM Guest` -> `Virtio`
- Useful internal libraries such as `posix-event`, `posix-futex` must be enabled.

Author: Good point to mention it here. The PR is already changing a lot of the config defaults with `imply` options, but you are right 9pfs is not covered if "Virtio-PCI" is not enabled and paging must (unfortunately) be enabled by hand (`imply` does not work if KVM wasn't enabled). I'll mention these settings :+1:

The resulting artifacts are available inside the `build/` directory (e.g. `build/app-elfloader_kvm-x86_64`).

> Under `example/helloworld` you can find an example application that you can use with the ELF Loader app. Just type `make` to compile it.

Reviewer: ```suggestion
Under `example/helloworld` you can find an example application that you can use with the ELF Loader app.
Type `make` to build it.
```

Reviewer: Mention that this builds two executables: a dynamic PIE executable and a static PIE executable.

Author: Good point. :+1:

Please refer to [Getting Started with Unikraft](https://unikraft.org/docs/getting-started/) for further details about the Unikraft build process and the available configuration options.

## Executing ELF binaries

> The `elfloader` currently supports statically linked and dynamically linked applications for Linux on x86_64 as long as they are compiled position independent (PIE).

Reviewer: ```suggestion
The `elfloader` currently supports statically-linked and dynamically-linked applications for Linux on x86_64, as long as they are compiled position independent (PIE).
```

Author: :+1:

For example, the application binary can be packaged with a CPIO initramdisk or handed over via a 9pfs host share.
Please note that we use 9pfs in the following how-to.
To load the application from another file system (e.g., initramdisk), you will need to follow equivalent steps.

Before we can launch an application we need to prepare a root file system that contains the ELF binary along with its library dependencies.
You can use `ldd` (or probably `musl-ldd` for applications linked with [`musl`](https://www.musl-libc.org/)) to list the shared libraries on which the application depends.
Please note that the vDSO (here: `linux-vdso.so.1`) is a kernel-provided library that is not present on the host filesystem.
Please ignore this file.

For a helloworld example application (here: [`/example/helloworld`](./example/helloworld), compiled on Debian 11), `ldd` will likely look like the following:

```sh
$ ldd helloworld
	linux-vdso.so.1 (0x00007ffdd695d000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007efed259f000)
	/lib64/ld-linux-x86-64.so.2 (0x00007efed2787000)
```

Copy the library dependencies to the same subdirectories as reported by `ldd`.

> Please remember to also copy any additional and required configuration file to the root file system. 

Reviewer: ```suggestion
Please remember to also copy any additional and required configuration files to the root file system. 
```

Author: :+1:

In this example, the populated root filesystem will look like this:

> ```

Reviewer: ```suggestion
```text
```

Author: Isn't the default `text` already?

rootfs/
├── lib
│   └── x86_64-linux-gnu
│       └── libc.so.6
├── lib64
│   └── ld-linux-x86-64.so.2
└── helloworld
```

Because the official dynamic loader maps the application and libraries into memory, the `elfloader` unikernel must be configured with `posix-mmap`, `ukvmem`, and `vfscore`.
For 9pfs, also make sure that you configured `vfscore` to automatically mount a host shared filesystem: Under `Library Configuration -> vfscore: Configuration` select `Automatically mount a root filesystem`, set `Default root filesystem` to `9PFS`, and ensure that `Default root device` is to `fs0`.
This last option simplifies the use of the `-e` parameter of [`qemu-guest`](https://github.com/unikraft/unikraft/tree/staging/support/scripts).

The application can then be started with:

```sh
> qemu-guest -k elfloader_kvm-x86_64 -e rootfs/ \

>             -a "/helloworld <application arguments>"

Reviewer: For me, this didn't work. The command that worked for me was:

```sh
../../unikraft/support/scripts/qemu-guest -k build/app-elfloader_kvm-x86_64 -e $(pwd)/rootfs/ -i $(pwd)/rootfs/lib64/ld-linux-x86-64.so.2 -a "/helloworld"
```

Reviewer: Add note that the application is going to print `helloworld` and then "hang". Use `Ctrl+c` to close it.

Reviewer: There should be an example of `helloworld` as a dynamic PIE and one with `helloworld` as a static PIE.

Reviewer: Running static PIE:

```
../../unikraft/support/scripts/qemu-guest -k build/app-elfloader_kvm-x86_64 -e $(pwd)/rootfs/ -i $(pwd)/rootfs/helloworld_static
```

Author: Please note the updated PR dependency of this PR. The setting should also be "Execute ELF from VFS", I'll double-check that this is mentioned.

> For me, this didn't work. The command that worked for me was:
> 
> ```shell
> ../../unikraft/support/scripts/qemu-guest -k build/app-elfloader_kvm-x86_64 -e $(pwd)/rootfs/ -i $(pwd)/rootfs/lib64/ld-linux-x86-64.so.2 -a "/helloworld"
> ```

Author: > Add note that the application is going to print `helloworld` and then "hang". Use `Ctrl+c` to close it.

:+1:

```

> NOTE: This assumes you built your unikernel with `APPELFLOADER_CUSTOMAPPNAME`.

Reviewer: This is not clear. I would remove it altogether.

Author: I prefer to mention which menuconfig option is expected to be set. It is (like the others previously mentioned) an `imply` option - meaning it could be unselected by mistake.

HINT: Environment variables can be set through `lib/posix-environ` and `lib/uklibparam`.
For this purpose, enable `Library Configuration -> posix-environ` and activate `Parse kernel command line arguments`.
The variables can be handed over via the kernel command line with the Unikraft library parameter `env.vars`, for example:

```sh
> qemu-guest -k elfloader_kvm-x86_64 -e rootfs/ \
            -a "env.vars=[ LD_LIBRARY_PATH=/lib LD_SHOW_AUXV=1 ] -- /helloworld <application arguments>"
```

## Debugging

### `strace`-like Output

> Unikraft's [`syscall_shim`](https://github.com/unikraft/unikraft/tree/staging/lib/syscall_shim) provides the ability to print a strace-like message for every processed binary system call request on the kernel output. This option can be useful for understanding what code a system call handler returns to the application and how the application interacts with the kernel.

Reviewer: ```suggestion
Unikraft's [`syscall_shim`](https://github.com/unikraft/unikraft/tree/staging/lib/syscall_shim) provides the ability to print a strace-like message for every processed binary system call request on the kernel output.
This option can be useful for understanding what code a system call handler returns to the application, and how the application interacts with the kernel.
```

Author: :+1: ☺︎

The setting can be found under `Library Configuration -> syscall_shim -> Debugging`: `'strace'-like messages for binary system calls`.

### GNU Debugger (gdb)

> It is possible to debug `elfloader` together with the loaded application and use the full set of debugging facilities for kernel and application at the same time.

Reviewer: ```suggestion
It is possible to debug `elfloader` together with the loaded application, and use the full set of debugging facilities for kernel and application at the same time.
```

Author: :+1:

In principle, `gdb` must only be made aware of the runtime memory layout of `elfloader` with the loaded application.
Thanks to the single address space layout, we gain easy debugability and full transparency.

> #### Static-PIE ELF executable

Reviewer: ```suggestion
#### Static-PIE ELF
```

Author: :+1:

> As a first step, `gdb` is started with loading the symbols from the `dbg` image of the `elfloader`. We map the symbols of ELF application with the gdb command `add-symbol-file` by specifying the application (with debug symbols) and the base load address.

Reviewer: ```suggestion
As a first step, `gdb` is started with loading the symbols from the `dbg` image of the `elfloader`.
We map the symbols of ELF application with the gdb command `add-symbol-file` by specifying the application (with debug symbols) and the base load address.
```

Author: 👍🏻

If `info` messages are enabled in `ukdebug`, this base address will be messaged by the loader like this:

```
ELF program loaded to 0x400101000-0x4001c2a08 (793096 B), entry at 0x40010ad50
```

To this address (here: `0x400101000`) you have to add the offset of the `.text` segment.
You can use `readelf -S` to find it out. In our example it is `0x92a0` (output shortened):

```
$ readelf -S helloworld_static
Section Headers:
  [Nr] Name              Type             Address           Offset
       Size              EntSize          Flags  Link  Info  Align
  [11] .text             PROGBITS         00000000000092a0  000092a0
       0000000000086eb0  0000000000000000  AX       0     0     32
```

The resulting address here is `0x40010a2a0`. The symbols of the static helloworld program can then be loaded from `gdb` with the following command:

```
(gdb) add-symbol-file -readnow helloworld_static 0x40010A2A0
```

> From this point you have symbol resolution in your debugger for both, the Unikraft elfloader and the loaded application.

Reviewer: ```suggestion
From this point you have symbol resolution in your debugger, for both the Unikraft elfloader and the loaded application.
```

Author: :+1:

NOTE: You can only set regular breakpoints within the application (`break` with GDB) after it got loaded into memory by elfloader (otherwise they will be ignored).

> The recommended procedure is 1) to set a breakpoint just after the application was loaded (e.g., the first system call that the application executes), 2) let the execution continue until the breakpoint is reached, and 3) set the interesting breakpoints within application space.

Reviewer: ```suggestion
The recommended procedure is:

1. Set a breakpoint just after the application was loaded (e.g., the first system call that the application executes).
1. Let the execution continue until the breakpoint is reached.
1. Set the interesting breakpoints within application space.
```

Author: :+1:

> #### Dynamically linked ELF executable

Reviewer: ```suggestion
#### Dynamically-linked ELF
```

Author: :+1:

Author: I'll keep executable because we kept it also for `Static-PIE ELF executable`

The principle of runtime address space layout for dynamically linked executables is the same as for statically linked executables.
The differences are that we have additionally loaded a dynamic loader together with the application and we have to load the symbols of each dependent dynamic library as well.

> For this purpose, we recommend to enable `strace`-like output in `syscall_shim` (read subsection: `strace`-like output).

Reviewer: ```suggestion
For this purpose, we recommend to enable `strace`-like output in `syscall_shim` (read subsection: [`strace`-like output](#strace-like-output)).
```

Author: :+1:

> It is the dynamic loader that will for each library 1) open the library, 2) parse its ELF header, 3) memory-map all needed sections into memory, and 4) close the library file again.

Reviewer: ```suggestion
It is the dynamic loader that will, for each library:
1. Open the library.
1. Parse its ELF header.
1. Memory-map all needed sections into memory.
1. Close the library file again.
```

Author: :+1:

#       `qemu-guest`).
# NOTE: Please note that regular breakpoints within the application or a shared
#       library can only be set after they have been loaded into memory. Usually
#       this is done by elfloader for the application and dynamic loader or by
#       the dynamic loader for shared libraries.
# HINT: You can use the `directory` command to specify additional paths that
#       `gdb` will use to search for source files.
#       For example, if you run your dynamically linked application with
#       Debian's libc, you can install (`apt install glibc-source`) and
#       extract the glibc sources under /usr/src/glibc.
#         --eval-command="directory /usr/src/glibc/glibc-2.31"
exec gdb \
	--eval-command="target remote $GDBSRV" \
	--eval-command="hbreak _ukplat_entry" \
	--eval-command="continue" \
	--eval-command="disconnect" \
	--eval-command="set arch i386:x86-64:intel" \
	--eval-command="target remote $GDBSRV" \
	\
	--eval-command="$( gdb-add-symbols "rootfs/helloworld" "8000000000" )" \
	--eval-command="$( gdb-add-symbols "rootfs/libc.so.6" "8000005000" )"
```

#### Hint: Debug symbols of libraries installed from packages

> If you run your dynamically linked application with libraries installed via a package manager, you can check if a debug package is also available for installation.

Reviewer: ```suggestion
If you run your dynamically-linked application with libraries installed via a package manager, you can check if a debug package is also available for installation.
```

Author: :+1:

For example, Debian provides the debug symbols automatically for `gdb` with the following installation:

```sh

> $ apt install libc6-dbg

Reviewer: ```suggestion
$ sudo apt install libc6-dbg
```

Author: On minimal Debian installations, `sudo` is not available. I'd expect, Debian users would understand what to do if `apt install` does not work. I could mention that the commands need to run as root and mention `sudo` in the text above. What do you think?

Reviewer: Yes, mention the need for install privileges. Or use `#` instead of `$` for the prompt.

Executing `make` a second time triggers a build of the unikernel for the
selected platform(s). The resulting artifacts are available inside the `build/`
directory (e.g. `build/app-elfloader_kvm-x86_64`).

Under `example/helloworld` you can find an example application that you can
use with the ELF Loader app. Just type `make` to compile it.

Please refer to [Getting Started with
Unikraft](https://unikraft.org/docs/getting-started/) for further details about
the Unikraft build process and the available configuration options.

## Executing ELF binaries

`elfloader` currently supports statically linked position-independent (PIE)
executables compiled for Linux on x86_64. Dynamically linked PIE executables can
be chain loaded using the corresponding official dynamic loader
(e.g., `ld-linux-x86-64.so.2` from glibc). This loader is also recognized as a
statically linked PIE executable.

In this version, any executable can only be passed to an `elfloader` unikernel
via the initrd argument. The ELF executable is the initrd, which means that no
additional initrd can be used as the root filesystem. Any additional filesystem
content must be passed via 9pfs.

> ### Static-PIE ELF executable
> A static PIE executable can be simply handed over as initrd

Reviewer: ```suggestion
### Static-PIE ELF executable

A static PIE executable can be simply handed over as initrd
```

> ([`qemu-guest`](https://github.com/unikraft/unikraft/tree/staging/support/scripts)):
> ```sh

Reviewer: ```suggestion
([`qemu-guest`](https://github.com/unikraft/unikraft/tree/staging/support/scripts)):

```sh
```

$ qemu-guest -k elfloader_kvm-x86_64 -i /path/to/elfprogram \
             -a "<application arguments>"

> ```
> Any application arguments are passed via kernel argument to the application.

Reviewer: ```suggestion
```

Any application arguments are passed via kernel argument to the application.
```

Environment variables cannot be set, currently.

> ### Dynamically linked ELF executable
> A dynamically linked PIE executable must be placed in a 9pfs root file system

Reviewer: ```suggestion
### Dynamically linked ELF executable

A dynamically linked PIE executable must be placed in a 9pfs root file system
```

	# Compute offset of .text section with base address
	LOAD_TADDR=$( printf 'obase=16;ibase=16;%s+%s\n' "${LOAD_ADDR^^}" "${TEXT_OFFSET^^}" | bc )

	# Generate GDB command
	printf 'add-symbol-file -readnow %s 0x%s' "${LOAD_ELF}" "${LOAD_TADDR}"
}

# Connect to $GDBSRV and set up gdb
# NOTE: The first block of instructions connects to the gdb port of the qemu
#       process and follows the CPU mode change while the guest is booting.
#       This is currently a requirement for using gdb with qemu for x86_64. For
#       other platforms and architectures, this may be different.
#       This block assumes that the guest is started in paused state (`-P` for
#       `qemu-guest`).
# HINT: You can use the `directory` command to specify additional paths that
#       `gdb` will use to search for source files.
#       For example, if you run your dynamically linked application with
#       Debian's libc, you can install (`apt install glibc-source`) and
#       extract the glibc sources under /usr/src/glibc.
#         --eval-command="directory /usr/src/glibc/glibc-2.31"
exec gdb \
	--eval-command="target remote $GDBSRV" \

> 	--eval-command="hbreak _libkvmplat_entry" \

Reviewer 2: This symbol name was removed in https://github.com/unikraft/unikraft/pull/722.

Probably it should be changed to `_multiboot_entry` or `_ukplat_entry`.

Author: Right, it got changed upstream in the meantime. I'll fix it. Thanks 👍🏻

Reviewer: What @csvancea said.

> # Executing ELF binaries

Reviewer 3: The commit title should be similar to the one about debugging, to mention an action e.g. Add section about executing. And should add as well what is executed / debugged e.g. Add section about executing ELF binaries ( with "debugging" instead of "executing", for the second commit).

Author: I made the commit titles looking similar. Since this is the app repository of `elfloader`, I'd prefer a shorter, more general title, like "Readme: Section about debugging". I think this is true because we are also executing and debugging the `elfloader` unikernel (and not only an ELF binary). What do you think?

Reviewer 3: Thank you for the overall updates. :) Indeed, it looks fine to be more general.

> `elfloader` currently supports statically linked position-independent (PIE)

Reviewer 3: Could add a section with the step(s) needed to build the "elfloader" unikernel. And, for example, can also include info on the arch / platform support for the unikernel, mention possible customization using the provided Makefile and the included options (e.g. " WITH_* ").

Author: I included @antoineco patch which should address this.

statically linked PIE executable.

In this version, any executable can only be passed to an `elfloader` unikernel
via the initrd argument. The ELF executable is the initrd, which means that no
additional initrd can be used as the root filesystem. Any additional filesystem
content must be passed via 9pfs.

## Static-PIE ELF executable
A static PIE executable can be simply handed over as initrd
([`qemu-guest`](https://github.com/unikraft/unikraft/tree/staging/support/scripts)):
```sh
$ qemu-guest -k elfloader_kvm-x86_64 -i /path/to/elfprogram \
             -a "<application arguments>"
```
Any application arguments are passed via kernel argument to the application.
Environment variables cannot be set, currently.

## Dynamically linked ELF executable
A dynamically linked PIE executable must be placed in a 9pfs root file system
along with its library dependencies. You can use `ldd` to list the dynamic
libraries on which the application depends in order to start.
Please note that the VDSO (here: `linux-vdso.so.1`) is provided by the Linux
kernel and is not present on the host filesystem. Please ignore this file.
```sh

> $ ldd helloworld

Reviewer 3: Could add a link reference to the hello world app location (https://github.com/unikraft/app-elfloader/tree/staging/example/helloworld), if using this one as an example. Just to have an intro about the config / build steps for the app that is used as an example (either statically or dynamically linked).

Author: I included @antoineco patch which should address this. I also added building of both versions of hello world as another commit to this PR: dynamically and statically linked.

	linux-vdso.so.1 (0x00007ffdd695d000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007efed259f000)
	/lib64/ld-linux-x86-64.so.2 (0x00007efed2787000)
```

A populated root filesystem may look like:
```
rootfs/
├── libc.so.6
└── helloworld
```

Because the official dynamic loader maps the application and libraries into

> memory, the `elfloader` unikernel must be configured with `posix-mmap` or

Reviewer 3: Is it possible to include these options in a config that can be used out of the box to build the "elfloader" unikernel? It can  be an alternative to setting them one by one.

Reviewer 4: > unikernel must be configured with `posix-mmap` or
> `ukmmap`.

I just tried building the app with plain defaults, and I see that `CONFIG_LIBUKMMAP=y` is set after running `make menuconfig` for the first time. This means that `ukmap` is already included and the user doesn't have to do anything if I'm not mistaken?

Author: In fact, elfloader can be built without ukmmap or posix-mmap. Config.uk reflects the most minimal configuration. At the moment, a kraft yaml file could include additional (optional) dependencies for a more meaningful build.

([`qemu-guest`](https://github.com/unikraft/unikraft/tree/staging/support/scripts)):
```sh
$ qemu-guest -k elfloader_kvm-x86_64 -i /path/to/elfprogram \
             -a "<application arguments>"
```
Any application arguments are passed via kernel argument to the application.
Environment variables cannot be set, currently.

### Dynamically linked ELF executable
A dynamically linked PIE executable must be placed in a 9pfs root file system
along with its library dependencies. You can use `ldd` to list the dynamic
libraries on which the application depends in order to start.
Please note that the VDSO (here: `linux-vdso.so.1`) is provided by the Linux
kernel and is not present on the host filesystem. Please ignore this file.
```sh
$ ldd helloworld
	linux-vdso.so.1 (0x00007ffdd695d000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007efed259f000)
	/lib64/ld-linux-x86-64.so.2 (0x00007efed2787000)
```

A populated root filesystem may look like:
```
rootfs/
├── libc.so.6
└── helloworld
```

Because the official dynamic loader maps the application and libraries into
memory, the `elfloader` unikernel must be configured with `posix-mmap` or

> `ukmmap`. In addition, you must include `vfscore` in the build and do the

Reviewer 3: The setup doesn't work with ukmmap, so only posix-mmap & ukvmem should be mentioned here.

Author: Sorry, you are right. `ukmmap` has still a point for being in for memory allocations for the application (if not done via `brk`/`sbrk`) but this has nothing to do with the support that the dynamic linker needs which is mapping files into memory. I will update this note. 👍🏻

following configuration: Under `Library Configuration -> vfscore: Configuration`
select `Automatically mount a root filesystem`, set `Default root filesystem` to
`9PFS`, and optionally set `Default root device` to `fs0`. This last option
simplifies the use of the `-e` parameter of `qemu-guest`.

A dynamically linked application (here: [`/helloworld`](./example/helloworld))
can then be started with
([`qemu-guest`](https://github.com/unikraft/unikraft/tree/staging/support/scripts)):
```sh
> qemu-guest -k elfloader_kvm-x86_64 -i /lib64/ld-linux-x86-64.so.2 -e rootfs/ \

>             -a "--library-path / /helloworld <application arguments>"

Reviewer 3: For the current setup, "ld-linux-x86-64.so.2" needs to be included in the command e.g. -a "/ld-linux-x86-64.so.2 --library-path ...". At least, from what I've checked out so far, it doesn't work without it.

Author: Thanks for confirming, right. I will update it 👍🏻

Author: Ouh... wait. This is actually only true if `APPELFLOADER_CUSTOMAPPNAME` is set.

Author: I assume that it is normally not set so the image name is taken as application name. I'll revert this single change and add a comment about this config option.

Please note that the VDSO (here: `linux-vdso.so.1`) is provided by the Linux
kernel and is not present on the host filesystem. Please ignore this file.
```sh
$ ldd helloworld
	linux-vdso.so.1 (0x00007ffdd695d000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007efed259f000)
	/lib64/ld-linux-x86-64.so.2 (0x00007efed2787000)
```

A populated root filesystem may look like:
```
rootfs/
├── libc.so.6
└── helloworld
```

Because the official dynamic loader maps the application and libraries into
memory, the `elfloader` unikernel must be configured with `posix-mmap`,
`ukvmem`, and `vfscore`. At the moment, you must configure `vfscore` to mount a
host shared filesystem: Under `Library Configuration -> vfscore: Configuration`
select `Automatically mount a root filesystem`, set `Default root filesystem` to
`9PFS`, and optionally set `Default root device` to `fs0`. This last option
simplifies the use of the `-e` parameter of `qemu-guest`.

A dynamically linked application (here: [`/helloworld`](./example/helloworld))
can then be started with
([`qemu-guest`](https://github.com/unikraft/unikraft/tree/staging/support/scripts)):
```sh

> > qemu-guest -k elfloader_kvm-x86_64 -i /lib64/ld-linux-x86-64.so.2 -e rootfs/ \

Reviewer 3: Is this command going to work with a follow-up fix / functionality update for the app-elfloader?

With the current setup and the "APPELFLOADER_CUSTOMAPPNAME" not set, the following error appears when trying to launch a VM using " -a "--library-path / /helloworld" " instead of " -a "/ld-linux-x86-64.so.2 --library-path / /helloworld" ".

```
/: error while loading shared libraries: /: cannot read file data: Error 21
```

Is still expected to push all args to the stack and update the following "for" loop, including "i>=0" instead of "i>0"?

https://github.com/unikraft/app-elfloader/blob/staging/elf_ctx.c#L172

Other than that, looks fine overall, thanks for the updates.

Reviewer 3: From what we discussed today, there is indeed the expectation to update the "for" loop and include "i>=0" instead of "i>0".

            -a "--library-path / /helloworld <application arguments>"
```
Please note that environment variables cannot be set, currently.

# Debugging

## `strace`-like Output

> Unikraft's `syscall_shim` provides the ability to print a strace-like message

Reviewer 3: Can add a reference to the codebase location ( https://github.com/unikraft/unikraft/tree/staging/lib/syscall_shim ) when mentioning "syscall_shim".

Author: I included @antoineco patch which should address this.

#       extract the glibc sources under /usr/src/glibc.
#         --eval-command="directory /usr/src/glibc/glibc-2.31"
exec gdb \
	--eval-command="target remote $GDBSRV" \
	--eval-command="hbreak _libkvmplat_entry" \
	--eval-command="continue" \
	--eval-command="disconnect" \
	--eval-command="set arch i386:x86-64:intel" \
	--eval-command="target remote $GDBSRV" \
	\
	--eval-command="$( gdb-add-symbols "rootfs/helloworld" "8000000000" )" \
	--eval-command="$( gdb-add-symbols "rootfs/libc.so.6" "8000005000" )"
```

### Hint: Debug symbols of libraries installed from packages

If you run your dynamically linked application with libraries installed via a
package manager, you can check if a debug package is also available for
installation.
For example, Debian provides the debug symbols automatically for `gdb` with the
following installation:
```sh
$ apt install libc6-dbg
```
As soon as the Debian's libc.so.6 is loaded by `gdb`, the debugger will load the
symbols provided by the Debian debug package.

### Hint: Sources of libraries installed from packages

If you run your dynamically linked application with libraries installed via a

> package manager, you can check if a source package available for installation.

Reviewer 3: "if a source package available" => "if a source package is available"

Author: 👍🏻
